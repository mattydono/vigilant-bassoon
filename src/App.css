.app-container {
    background-color: lightgrey;
    display: flex;
    justify-content: center;
    align-items: center;
    height: 600px;
    max-width: 60%;
    min-width: 300px;
    margin: 150px auto;
    border-radius: 10px;
    box-shadow: 0 0 20px #333;
}

.chat-container {
    display: flex;
    flex-direction: column;
    width: 100%;
    height: 100%;
}