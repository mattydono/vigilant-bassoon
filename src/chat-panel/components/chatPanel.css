.chat-container {
  height: 100%;
  width: 100%;
  max-height: 100%;
  display: flex;
  flex-direction: column;
}

.chatOutput-container {
  background-color: lightgrey;
  height: 70%;
  max-height: 70%;
  width: 100%;
  max-width: 100%;
  display: flex;
  flex-direction: column;
  overflow: auto;
}

.messageDiv {
  flex: 0 0 auto;
  min-height: 30px;
  width: 40%;
  max-width: 100%;
  display: flex;
  justify-content: space-between;
  background-color: darkorange;
  align-items: baseline;
  align-self: flex-start;
  overflow-wrap: break-word;
  word-wrap: break-word;
  hyphens: auto;
  white-space: pre-wrap;
  margin-top: 1%;
  border-radius: 5px;
}

.messageDivActive {
  flex: 0 0 auto;
  min-height: 30px;
  width: 40%;
  max-width: 100%;
  display: flex;
  justify-content: space-between;
  background-color: aquamarine;
  align-items: baseline;
  align-self: flex-end;
  overflow-wrap: break-word;
  word-wrap: break-word;
  hyphens: auto;
  white-space: pre-wrap;
  margin-top: 1%;
  border-radius: 5px;
  position: relative;
}

.messageDivActive:hover .messageEdit {
  opacity: 1;
}

.messageEdit {
  opacity: 0;
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
  transform: translate(-50%, -25%);
  transition: opacity 200ms ease-in-out;
  display: flex;
}

.editButton {
  border: none;
  outline: none;
  background-color: rgba(0, 0, 0, 0);
  padding: 0;
  border-radius: 5px;
  height: 20px;
  width: 20px;
  display: flex;
  flex-grow: 0;
  justify-content: center;
}

.editButton:hover {
  background-color: rgba(100, 185, 185, 1);
  cursor: pointer;
}

.chatInput-container {
  background-color: yellow;
  height: 30%;
  width: 100%;
  display: flex;
}

.input {
  resize: none;
  width: 100%;
  height: 100%;
  box-sizing: border-box;
}

.button {
  height: 100%;
  width: 20%;
  box-sizing: border-box;
  white-space: initial;
}

.scroll {
  height: 600px;
  width: 400px;
}

.xsFont {
  font-size: x-small;
  margin-left: 5%;
  margin-right: 5%;
}
