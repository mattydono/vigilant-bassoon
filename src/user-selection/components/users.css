.users-container {
  background-color: black;
  height: 100%;
  max-height: 100%;
  width: 30%;
  display: flex;
  flex-direction: column;
}

.userDisplay {
  height: 90%;
  max-height: 100%;
  width: 100%;
  background-color: yellow;
  overflow: auto;
}

.userOption {
  height: 50px;
  width: 100%;
  max-width: 100%;
  border: solid 1px black;
  border-bottom-width: 0px;
  display: flex;
  align-items: center;
  box-sizing: border-box;
  display: flex;
  flex-direction: row;
  justify-content: center;
}

.userOption:hover {
  cursor: pointer;
}

.active:hover .removeUser {
  display: block;
}

.userOption:last-child {
  border-bottom-width: 1px;
}

.inactive:hover {
  background-color: darkorange;
}

.active {
  background-color: aquamarine;
  border-right: dashed 1px black;
}

.removeUser {
  display: none;
  position: relative;
  right: 25%;
}

.inputContainer {
  height: 10%;
  width: 100%;
}

.userInput {
  height: 100%;
  width: 75%;
  box-sizing: border-box;
}

.userPush {
  height: 100%;
  width: 25%;
  background-color: white;
  box-sizing: border-box;
}
