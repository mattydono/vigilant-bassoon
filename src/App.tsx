import React, { Component } from 'react';
import uuid from 'uuid/v4';
import './App.css';
import { ChatPanel, Message } from './chat-panel';
import { User, Users } from './user-selection';

const getCurrentTime = () => new Date().toLocaleTimeString();

type State = {
  users: User[];
  activeUserId: string | null;
  messages: Message[];
};

export class App extends Component<{}, State> {
  public state: State = {
    users: [],
    activeUserId: null,
    messages: [],
  };

  public render() {
    const activeUser = this.state.users.find(user => user.id === this.state.activeUserId);

    return (
      <div className="app-container">
        <Users
          users={this.state.users}
          activeUserId={this.state.activeUserId}
          setActiveUser={this.setActiveUser}
          addUser={this.addUser}
        />
        {activeUser !== undefined && (
          <ChatPanel
            messages={this.state.messages}
            activeUserName={activeUser.name}
            addMessage={this.addMessage}
          />
        )}
      </div>
    );
  }

  private setActiveUser = (activeUserId: string | null) => {
    this.setState({ activeUserId });
  };

  private addUser = (name: string) => {
    this.setState(state => ({
      users: [...state.users, { id: uuid(), name }],
    }));
  };

  private addMessage = (message: string) => {
    const activeUser = this.state.users.find(user => {
      return user.id === this.state.activeUserId;
    });

    if (activeUser !== undefined) {
      this.setState(state => ({
        messages: [...state.messages, { id: activeUser.name, message, time: getCurrentTime() }],
      }));
    }
  };
}
