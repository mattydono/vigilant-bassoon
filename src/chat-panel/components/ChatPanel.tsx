import React, { Component, createRef } from 'react';
import { connect } from 'react-redux';
import uuid from 'uuid';
import { AppState } from '../../redux/configureStore';
import { User } from '../../user-selection';
import { Message, MessageId } from '../Message';
import { addMessage, removeMessage } from '../redux';
import './chatPanel.css';

type State = {
  userText: string;
};

type ChatPanelMessage = Message & { user: User };

type StateProps = {
  messages: ChatPanelMessage[];
  activeUser: User | null;
};

type DispatchProps = {
  addMessage: typeof addMessage; // both the same
  removeMessage: (messageId: MessageId) => void;
};

type Props = StateProps & DispatchProps;

class _ChatPanel extends Component<Props, State> {
  public state: State = {
    userText: '',
  };

  private scrollContainer = createRef<HTMLDivElement>();

  public componentDidUpdate(prevProps: Props) {
    if (this.props.messages.length !== prevProps.messages.length) {
      this.scrollToBottom();
    }
  }

  public render() {
    if (!this.props.activeUser) {
      return null;
    }

    return (
      <div className="chat-container">
        <div className="chatOutput-container" ref={this.scrollContainer}>
          {this.renderMessage()}
        </div>
        <div className="chatInput-container">
          <textarea
            className="input"
            placeholder="Text Here..."
            value={this.state.userText}
            onChange={this.onTextUpdate}
            onKeyPress={this.onKeyPress}
          />
          <button className="button" onClick={this.onTextSubmit}>
            Enter
          </button>
        </div>
      </div>
    );
  }

  private onKeyPress: React.KeyboardEventHandler<HTMLTextAreaElement> = e => {
    if (e.key === 'Enter' && !e.shiftKey) {
      e.preventDefault();
      this.onTextSubmit();
    }
  };

  private scrollToBottom = () => {
    const container = this.scrollContainer.current;
    if (container !== null) {
      container.scrollTop = container.scrollHeight;
    }
  };

  private renderMessage = () =>
    this.props.messages.map(message => (
      <div
        className={this.props.activeUser!.id === message.userId ? 'messageDivActive' : 'messageDiv'}
      >
        <div className="messageEdit" onClick={this.onMessageDeletion(message.id)}>
          ❌✎
        </div>
        <div className="xsFont">{message.user.name}:</div>
        <div>{message.message}</div>
        <div className="xsFont">{message.time}</div>
      </div>
    ));

  private onMessageDeletion = (messageId: MessageId) => () => {
    this.props.removeMessage(messageId);
  };

  private onTextUpdate: React.ChangeEventHandler<HTMLTextAreaElement> = event => {
    const { value: userText } = event.currentTarget;
    this.setState(() => ({ userText }));
  };

  private onTextSubmit = () => {
    const message = this.state.userText;
    if (message === '') {
      return;
    }
    this.props.addMessage({
      userId: this.props.activeUser!.id,
      message,
      id: uuid(),
      time: new Date().toLocaleTimeString(),
    });
    this.setState(() => ({ userText: '' }));
  };
}

function mapStateToProps(state: AppState): StateProps {
  const messages = state.messages.messages.map<ChatPanelMessage>(message => {
    const user = state.users.users.find(user => user.id === message.userId);
    if (user === undefined) {
      return { ...message, user: { id: message.userId, name: 'DELETED' } };
    }
    return { ...message, user };
  });

  const activeUser = state.users.users.find(user => user.id === state.users.activeUserId);
  return {
    messages,
    activeUser: activeUser ? activeUser : null,
  };
}

// function mapDispatchToProps(dispatch: Dispatch): DispatchProps {
//   return {
//       addMessage: (message: Message) => dispatch(addMessage(message)),
//       removeMessage: (messageId: MessageId) => dispatch(removeMessage(messageId)),
//   }
// }

const dispatchMap: DispatchProps = {
  addMessage,
  removeMessage,
};

export const ChatPanel = connect(
  mapStateToProps,
  dispatchMap,
)(_ChatPanel);
