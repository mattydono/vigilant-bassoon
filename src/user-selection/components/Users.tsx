import React from 'react';
import { User } from '../User';
import './users.css';

type Props = {
  users: User[];
  activeUserId: string | null;
  setActiveUser: (id: string | null) => void;
  addUser: (name: string) => void;
};

type State = {
  name: string;
};

export class Users extends React.Component<Props, State> {
  public state: State = {
    name: '',
  };

  private scrollContainer = React.createRef<HTMLDivElement>();

  public componentDidUpdate(prevProps: Props) {
    if (this.props.users.length !== prevProps.users.length) {
      this.scrollToBottom();
    }
  }

  public render() {
    return (
      <div className="users-container">
        <div className="userDisplay" ref={this.scrollContainer}>
          {this.renderedUser()}
        </div>
        <div className="inputContainer">
          <input
            className="userInput"
            placeholder="Enter Name"
            value={this.state.name}
            onChange={this.onNameChange}
            onKeyPress={this.onKeyPress}
          />
          <button className="userPush" onClick={this.onClickAdd}>
            Enter
          </button>
        </div>
      </div>
    );
  }

  private renderedUser = () => {
    const { users, activeUserId } = this.props;
    return users.map(user => {
      const className = activeUserId === user.id ? 'active' : 'inactive';
      const onClick = this.userSelect(activeUserId === user.id ? null : user.id);
      return (
        <div key={user.id} className={`userOption ${className}`} onClick={onClick}>
          {user.name}
        </div>
      );
    });
  };

  private onKeyPress: React.KeyboardEventHandler<HTMLInputElement> = event => {
    if (event.key === 'Enter') {
      this.onClickAdd();
    }
  };

  private onNameChange: React.ChangeEventHandler<HTMLInputElement> = event => {
    const { value: name } = event.currentTarget;
    this.setState(() => ({ name }));
  };

  private nameExists = (name: string) => this.props.users.some(user => user.name === name);

  private onClickAdd = () => {
    const name = this.state.name.trim();
    if (name.length === 0 || this.nameExists(name)) {
      return;
    }

    this.props.addUser(name);
    this.clear();
  };

  private clear = () => {
    this.setState({ name: '' });
  };

  private userSelect = (id: string | null) => () => {
    this.props.setActiveUser(id);
  };

  private scrollToBottom = () => {
    const container = this.scrollContainer.current;
    if (container !== null) {
      container.scrollTop = container.scrollHeight;
    }
  };
}
