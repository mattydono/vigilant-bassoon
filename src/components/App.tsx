import React, {Component} from 'react';
import {Users} from './users/Users';
import {ChatInput} from './chat_input/ChatInput';
import '../app.css';
import {Message, User} from "../models";
import uuid from 'uuid/v4';

type State = {
    users: User[];
    activeUserId: string | null;
    messages: Message[];
};

// const USERS: ReadonlyArray<User> = [
//     {id: uuid(), name: 'Kyle'},
//     {id: uuid(), name: 'Matt'},
// ];

export class App extends Component<{}, State> {
    public state: State = {
        users: [],
        activeUserId: null,
        messages: [],
    };

    render() {

        const activeUser = this.state.users.find(user => {
           return user.id === this.state.activeUserId
        });

        return (
            <div className="app-container">
                <Users users={this.state.users} activeUserId={this.state.activeUserId}
                       setActiveUser={this.setActiveUser} addUser={this.addUser}/>
                {activeUser !== undefined && <ChatInput messages={this.state.messages} activeUserName={activeUser.name} addMessage={this.addMessage}/>}
            </div>
        );
    }

    private setActiveUser = (activeUserId: string | null) => {
        this.setState({activeUserId});
    };

    private addUser = (name: string) => {
        this.setState ((state) => ({
            users: [...state.users, {name, id: uuid()}]
        }));
    };

    private addMessage = (message: string) => {

        const activeUser = this.state.users.find(user => {
            return user.id === this.state.activeUserId
        });

        if(activeUser !== undefined) {

            this.setState((state) => ({
                messages: [...state.messages, {
                    message,
                    id: activeUser.name,
                    time: this.getCurrentTime()
                }],
            }));
        }
    };
    private getCurrentTime = () => new Date().toLocaleTimeString();
}