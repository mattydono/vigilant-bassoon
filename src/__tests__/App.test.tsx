import { shallow } from 'enzyme';
import React from 'react';
import { App } from '../components/App';

describe('The App root', () => {
  it('renders as expected', () => {
    const shallowRendered = shallow(<App />);
    expect(shallowRendered).toMatchSnapshot();
  });
});
