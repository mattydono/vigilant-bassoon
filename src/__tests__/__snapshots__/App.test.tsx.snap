// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The App root renders as expected 1`] = `
<div
  className="App"
>
  <Input
    onChange={[Function]}
  />
  <Component
    content=""
  />
</div>
`;
